src/source_process.c

```diff
@@ -13,25 +13,38 @@
 #include <unistd.h>
 #include <stdio.h>
 
+#define MIN(a,b) ((a) < (b) ? (a) : (b))
+
+#define POW_CACHE_SPLIT 6
+#define POW_CACHE_DIVISOR (1 << POW_CACHE_SPLIT)
+
+
 static uint64_t m_for(uint64_t n) {
   return 2*((n+1)/4)+1;
 }
 
 typedef struct {
-  uint64_t n, n_args, m, p, //n = number of veritcies //obvious
+  uint64_t n, n_args, m, m_half, p, //n = number of veritcies //obvious
   p_dash, r, r2, // montgomery stuff, a _M suffix implies something is in montgomery form
+  *rs,
   *jk_prod_M, // cache of w^j*w^-k / (w^-j*w^k + w^j*w^-k)
   *nat_M, // natural numbers up to n (inclusive)
   *nat_inv_M, // inverses of natural numbers up to n (inclusive)
   *jk_sums_M, // w^-j*w^k + w^j*w^-k
+  *jk_sums_pow_upper_M, // see jk_sums_pow
+  *jk_sums_pow_lower_M,
   *ws_M, // powers of omega (m form)
   *fact_M, // i! for i <= n
   *fact_inv_M; // 1/i! for i <= n
 } prim_ctx_t;
 
-// index into rectangular array
+// We have caches of w^-j*w^k + w^j*w^-k and w^j*w^-k / (w^-j*w^k + w^j*w^-k)
+// Note that w^-j*w^k = w^(k-j)
+// So the value of these expressions is only actually a function of j-k. We use
+// this to make the caches smaller and cache more stuff whilst still keeping everying in L1
 static inline size_t jk_pos(size_t j, size_t k, uint64_t m) {
```

Comment:
As noted elsewhere we could shrink the cache size in half for the jk_sums cache. I don't think that's worth it here though as these caches are only large when there's an extra dimension.

What I am considering is we could do the opposite? Replicate the cache into the negative direction and store a pointer to the middle to remove the need to do the `result >= 0 ? (uint64_t) result : result + m;` operation.

I'm not really sure if that difference would even be big enough to profile? I'd welcome thoughts.